src/event/lookup/mod.rs

```diff
@@ -0,0 +1,274 @@
+mod segment;
+#[cfg(test)]
+mod test;
+
+use crate::{
+    event::Value,
+    mapping::parser::{MappingParser, Rule},
+};
+use pest::{iterators::Pair, Parser};
+use std::{
+    convert::TryFrom,
+    ops::{RangeFrom, RangeFull, RangeTo, RangeToInclusive},
+    slice::Iter,
+    str,
+};
+use toml::Value as TomlValue;
+
+use core::fmt;
+use indexmap::map::IndexMap;
+pub use segment::Segment;
+use serde::de::{self, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::fmt::{Display, Formatter};
+use std::ops::{Index, IndexMut};
+use std::str::FromStr;
+
+/// Lookups are pre-validated event lookup paths.
+///
+/// They are intended to handle user input from a configuration.
+///
+/// Generally, these shouldn't be created on the hot path.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+pub struct Lookup {
+    segments: Vec<Segment>,
+}
+
+impl<'a> TryFrom<Pair<'a, Rule>> for Lookup {
+    type Error = crate::Error;
+
+    fn try_from(pair: Pair<'a, Rule>) -> Result<Self, Self::Error> {
+        Ok(Self {
+            segments: Segment::from_lookup(pair)?,
+        })
+    }
+}
+
+impl Display for Lookup {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        let mut peeker = self.segments.iter().peekable();
+        let mut next = peeker.next();
+        let mut maybe_next = peeker.peek();
+        while let Some(segment) = next {
+            match segment {
+                Segment::Field(_) => {
+                    match maybe_next {
+                        Some(next) if next.is_field() => write!(f, r#"{}."#, segment)?,
+                        None | Some(_) => write!(f, "{}", segment)?,
+                    };
```

Comment:
rustfmt produce extra padding compare to next match because semicolon?

Reply:
Yeah! Can remove that!